File: be1-go/message/messagedata/federation_init.go
package messagedata

// FederationInit defines a message data
type FederationInit struct {
	Object string `json:"object"`
	Action string `json:"action"`

	LaoId         string `json:"lao_id"`
	ServerAddress string `json:"server_address"`
	PublicKey     string `json:"public_key"`
	ChallengeMsg  string `json:"challenge"`
}

// GetObject implements MessageData
func (FederationInit) GetObject() string {
	return FederationObject
}

// GetAction implements MessageData
func (FederationInit) GetAction() string {
	return FederationActionInit
}

// NewEmpty implements MessageData
func (FederationInit) NewEmpty() MessageData {
	return &FederationInit{}
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/federation/FederationInit.scala
package ch.epfl.pop.model.network.method.message.data.federation

import ch.epfl.pop.json.MessageDataProtocol.FederationInitFormat
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.Message
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, PublicKey}
import spray.json.*

final case class FederationInit(
    laoId: Hash,
    serverAddress: String,
    publicKey: PublicKey,
    challenge: Message
) extends MessageData {
  override val _object: ObjectType = ObjectType.federation
  override val action: ActionType = ActionType.init
}

object FederationInit extends Parsable {
  def apply(laoId: Hash, serverAddress: String, publicKey: PublicKey, challenge: Message): FederationInit = {
    new FederationInit(laoId, serverAddress, publicKey, challenge)
  }

  override def buildFromJson(payload: String): FederationInit = payload.parseJson.asJsObject.convertTo[FederationInit]
}
